File: src/main/scala/robosky/ether/block/UplanderFarmlandBlock.scala
package robosky.ether.block

import net.minecraft.block.{Block, BlockPlacementEnvironment, BlockState, Blocks}
import net.minecraft.block
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.{Entity, EntityContext, LivingEntity}
import net.minecraft.util.math.BlockPos
import net.minecraft.util.shape.VoxelShape
import net.minecraft.world.{BlockView, GameRules, World}


object UplanderFarmlandBlock {
    private val SHAPE: VoxelShape =
        Block.createCuboidShape(0.0D,
            0.0D,
            0.0D,
            16.0D,
            15.0D,
            16.0D
        )
}

class UplanderFarmlandBlock(val settings: Block.Settings) extends Block(settings) {
    override def getOutlineShape(state: BlockState, view: BlockView, pos: BlockPos, context: EntityContext): VoxelShape
        = UplanderFarmlandBlock.SHAPE

    override def canPlaceAtSide(blockState_1: BlockState, blockView_1: BlockView, blockPos_1: BlockPos,
                                blockPlacementEnvironment_1: BlockPlacementEnvironment): Boolean = false

    override def onLandedUpon (world_1: World, blockPos_1: BlockPos, entity_1: Entity, float_1: Float): Unit =  {
        if (!world_1.isClient && world_1.random.nextFloat < float_1 - 0.5F && entity_1.isInstanceOf[LivingEntity]
            && (entity_1.isInstanceOf[PlayerEntity] || world_1.getGameRules.getBoolean(GameRules.MOB_GRIEFING))
            && entity_1.getWidth * entity_1.getWidth * entity_1.getHeight > 0.512F) {
            setToDirt(world_1.getBlockState(blockPos_1), world_1, blockPos_1)
        }

        super.onLandedUpon(world_1, blockPos_1, entity_1, float_1)
    }

    def setToDirt(blockState_1: BlockState, world_1: World, blockPos_1: BlockPos) =
        world_1.setBlockState(blockPos_1, BlockRegistry.UPLANDER_DIRT.getDefaultState)
}

File: src/main/java/robosky/uplands/block/UplanderFarmlandBlock.java
package robosky.uplands.block;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.ShapeContext;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.ai.pathing.NavigationType;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.world.BlockView;
import net.minecraft.world.GameRules;
import net.minecraft.world.World;

public class UplanderFarmlandBlock extends Block {
    
    private static final VoxelShape SHAPE = Block.createCuboidShape(0.0D, 0.0D, 0.0D, 16.0D, 15.0D, 16.0D);
    
    public UplanderFarmlandBlock(Settings settings) {
        super(settings);
    }

    @Override
    public VoxelShape getOutlineShape(BlockState blockState, BlockView blockView, BlockPos blockPos, ShapeContext entityContext) {
        return SHAPE;
    }

    @Override
    public boolean canPathfindThrough(BlockState blockState, BlockView blockView, BlockPos blockPos, NavigationType blockPlacementEnvironment) {
        return false;
    }
    
    @Override
    public void onLandedUpon(World world, BlockPos blockPos, Entity entity, float f) {
        if (!world.isClient && world.random.nextFloat() < f - 0.5F && entity instanceof LivingEntity
                && (entity instanceof PlayerEntity || world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING))
                && entity.getWidth() * entity.getWidth() * entity.getHeight() > 0.512F) {

            setToDirt(world.getBlockState(blockPos), world, blockPos);
        }

        super.onLandedUpon(world, blockPos, entity, f);
    }

    private void setToDirt(BlockState blockState, World world, BlockPos blockPos) {
        world.setBlockState(blockPos, Block.pushEntitiesUpBeforeBlockChange(blockState, BlockRegistry.UPLANDER_DIRT.getDefaultState(), world, blockPos));
    }
}
